core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -47,43 +47,49 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
     with WorkspaceSupport
     with LazyLogging {
 
-  def createSnapshot(workspaceName: WorkspaceName, snapshot: NamedDataRepoSnapshot): Future[DataRepoSnapshotResource] =
+  def createSnapshot(workspaceName: WorkspaceName,
+                     snapshotIdentifiers: NamedDataRepoSnapshot
+  ): Future[DataRepoSnapshotResource] =
     getV2WorkspaceContextAndPermissions(workspaceName,
                                         SamWorkspaceActions.write,
                                         Some(WorkspaceAttributeSpecs(all = false))
-    ).flatMap { workspaceContext =>
-      val wsid = workspaceContext.workspaceIdAsUUID // to avoid UUID parsing multiple times
-      // create the stub workspace in WSM if it does not already exist
-      if (!workspaceStubExists(wsid, ctx)) {
-        workspaceManagerDAO.createWorkspace(wsid, workspaceContext.workspaceType, ctx)
+    ).flatMap { rawlsWorkspace =>
+      val wsid = rawlsWorkspace.workspaceIdAsUUID // to avoid UUID parsing multiple times
+      val snapshot =
+        new WrappedSnapshot(dataRepoDAO.getSnapshot(snapshotIdentifiers.snapshotId, ctx.userInfo.accessToken))
+      val snapshotValidator = new SnapshotReferenceCreationValidator(rawlsWorkspace, snapshot)
+      val aggregatedWorkspaceService = new AggregatedWorkspaceService(workspaceManagerDAO)
+
+      // prevent snapshots from disallowed platforms
+      snapshotValidator.validateSnapshotPlatform()
+
+      // prevent disallowed access across workspace or dataset protection boundaries
+      snapshotValidator.validateProtectedStatus()
+
+      try {
+        // if there's an existing WSM workspace, make sure its platform is compatible
+        // with that of the snapshot's dataset.
+        val wsmWorkspace = aggregatedWorkspaceService.fetchAggregatedWorkspace(rawlsWorkspace, ctx)
+        snapshotValidator.validateWorkspacePlatformCompatibility(wsmWorkspace.getCloudPlatform)
+      } catch {
+        case _: AggregateWorkspaceNotFoundException =>
+          // create the stub workspace in WSM if it does not already exist
+          workspaceManagerDAO.createWorkspace(wsid, rawlsWorkspace.workspaceType, ctx)
       }
```

Comment:
thinking out loud … 

If this is a GCP workspace, and if the stub workspace does not exist, `fetchAggregatedWorkspace` will throw `AggregateWorkspaceNotFoundException` and this code will never call `snapshotValidator.validateWorkspacePlatformCompatibility(…)`.

In our current business logic, this is fine, because we have already short-circuited if we have an Azure snapshot. I wonder if it this behavior could bite us in the future. I am not coming up with the use case, but it feels wiggly.

You could do something like:
```scala
      val cloudPlatform = try {
        // if there's an existing WSM workspace, make sure its platform is compatible
        // with that of the snapshot's dataset.
        val wsmWorkspace = aggregatedWorkspaceService.fetchAggregatedWorkspace(rawlsWorkspace, ctx)
        wsmWorkspace.getCloudPlatform
      } catch {
        case _: AggregateWorkspaceNotFoundException =>
          // create the stub workspace in WSM if it does not already exist
          workspaceManagerDAO.createWorkspace(wsid, rawlsWorkspace.workspaceType, ctx)
          WorkspaceCloudPlatform.Gcp
      }

      snapshotValidator.validateWorkspacePlatformCompatibility(cloudPlatform)
```